### PR TITLE
Add a Zephyr example.

### PR DESCRIPTION
**Pull request by haydenroche5**

### CODE REVIEW

File: examples/zephyr/main.c [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        strncpy(cache->valueA, val, sizeof(cache->valueA));
+    } else if (strcmp(var, "variable_b") == 0) {
+        strncpy(cache->valueB, val, sizeof(cache->valueB));
+    } else if (strcmp(var, "variable_c") == 0) {
+        strncpy(cache->valueC, val, sizeof(cache->valueC));
+    }
+
+    printk("variable_a has value %s\n", envVarCache.valueA);
+    printk("variable_b has value %s\n", envVarCache.valueB);
+    printk("variable_c has value %s\n", envVarCache.valueC);
```
Comment (zfields) [inline examples/zephyr/main.c:77]: We can make the log output easier to digest without requiring source code context.

```suggestion
    printk("\nCallback received variable (%s) with value (%s) and context (%p).\n", var, val, userCtx);

    /* Cache the values for each variable. */
    if (strcmp(var, "variable_a") == 0) {
        strncpy(cache->valueA, val, sizeof(cache->valueA));
    } else if (strcmp(var, "variable_b") == 0) {
        strncpy(cache->valueB, val, sizeof(cache->valueB));
    } else if (strcmp(var, "variable_c") == 0) {
        strncpy(cache->valueC, val, sizeof(cache->valueC));
    }

    /* Print cached values */
    printk("Cached values:\n");
    printk("- variable_a has value %s\n", envVarCache.valueA);
    printk("- variable_b has value %s\n", envVarCache.valueB);
    printk("- variable_c has value %s\n", envVarCache.valueC);
```

I looked at the output before I read the code, and the repetition was not immediately obvious...

```none
{"req":"env.get","names":["variable_a","variable_b","variable_c"]}
{"body":{}}
{"req":"env.get","names":["variable_a","variable_b","variable_c"]}
{"body":{"variable_a":"Good","variable_b":"job","variable_c":"Hayden!"}}
variable_a has value Good
variable_b has value 
variable_c has value 
variable_a has value Good
variable_b has value job
variable_c has value 
variable_a has value Good
variable_b has value job
variable_c has value Hayden!
{"req":"env.get","names":["variable_a","variable_b","variable_c"]}
{"body":{"variable_a":"Good","variable_b":"job","variable_c":"Hayden!"}}
variable_a has value Good
variable_b has value job
variable_c has value Hayden!
variable_a has value Good
variable_b has value job
variable_c has value Hayden!
variable_a has value Good
variable_b has value job
variable_c has value Hayden!
```

File: examples/zephyr/main.c [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    JAddStringToObject(req, "mode", "continuous");
+    JAddBoolToObject(req, "sync", true);
+    /*
+     * Send the request with a retry timeout. If the Notecard has just started
+     * up, it may need a moment before it's able to receive and respond to
+     * requests.
+     */
+    if (!NoteRequestWithRetry(req, HUB_SET_RETRY_SECONDS)) {
+        printk("hub.set failed, aborting.\n");
+        return;
```
Comment (zfields) [inline examples/zephyr/main.c:105]:
```suggestion
        return 0;
```

File: examples/zephyr/main.c [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    /* Allocate the environment variable manager. */
+    envVarManager = NotecardEnvVarManager_alloc();
+    if (envVarManager == NULL) {
+        printk("Failed to allocate env var manager.\n");
+        return;
```
Comment (zfields) [inline examples/zephyr/main.c:112]:
```suggestion
        return 0;
```

File: examples/zephyr/main.c [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    /*
+     * Set the callback for the manager, and give it a pointer to our cache
+     * so that we can store the values of the environment variables we're
+     * fetching.
+     */
+    if (NotecardEnvVarManager_setEnvVarCb(envVarManager, envVarManagerCb,
+                                          &envVarCache) != NEVM_SUCCESS) {
+        printk("Failed to set env var manager callback.\n");
+        return;
```
Comment (zfields) [inline examples/zephyr/main.c:123]:
```suggestion
        return 0;
```

**Review by zfields — CHANGES_REQUESTED** [review]
see comments

**Comment by zfields** [comment]
Arduino example compiles with warning...

```none
/home/zak/Development/Arduino/libraries/notecard-env-var-manager/examples/arduino/arduino.ino: In function 'void envVarManagerCb(const char*, const char*, void*)':
/home/zak/Development/Arduino/libraries/notecard-env-var-manager/examples/arduino/arduino.ino:40:16: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 16 equals destination size [-Wstringop-truncation]
   40 |         strncpy(cache->valueB, val, sizeof(cache->valueB));
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/zak/Development/Arduino/libraries/notecard-env-var-manager/examples/arduino/arduino.ino:37:16: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 16 equals destination size [-Wstringop-truncation]
   37 |         strncpy(cache->valueA, val, sizeof(cache->valueA));
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

We can clear the warning by substracting 1 from the `strncpy()` size.

Example:
```c
strncpy(cache->valueB, val, sizeof(cache->valueB)-1);
```

https://stackoverflow.com/questions/50198319/gcc-8-wstringop-truncation-what-is-the-good-practice

**Comment by haydenroche5** [comment]
> We also need to find a new home for the Zephyr example. The Arduino IDE will try to compile any folder it finds under `examples/` folder.
> 
> ```
> /home/zak/Development/GitHub/blues/notecard-env-var-manager/examples/zephyr/build/zephyr/misc/generated/syscalls_links/include_zephyr_linker_common-rom (No such file or directory)
> Invalid library found in /home/zak/Development/Arduino/libraries/notecard-env-var-manager: no headers files (.h) found in /home/zak/Development/Arduino/libraries/notecard-env-var-manager
> ```

Not seeing that on my end. I launch Arduino IDE, open examples/arduino/arduino.ino, and it builds without errors. Are you doing something different from that?

**Review by zfields — APPROVED** [review]

**Pull request merged**